### ISSUE TITLE
HW 6 Plots and what not

### ISSUE THREAD

**Issue by kaiyun717** [kaiyun717/machine-learning#1]
# 2(b) FINAL
```python
model = nn.Sequential(
    nn.Flatten(),
    nn.Linear(28*28, 128),
    nn.ReLU(),
    nn.Dropout(p=0.30),
    nn.Linear(128, 64),
    nn.ReLU(),
    nn.Linear(64, 10)
)

batch_size = 100
n_epochs = 20
lr = 5e-3 
```

```console
Train Epoch: 1  Average loss: 0.0034
Valid set: Average loss: 0.0018, Accuracy: 11362/12000 (94.6833)

Train Epoch: 2  Average loss: 0.0019
Valid set: Average loss: 0.0013, Accuracy: 11538/12000 (96.1500)

Train Epoch: 3  Average loss: 0.0015
Valid set: Average loss: 0.0013, Accuracy: 11544/12000 (96.2000)

Train Epoch: 4  Average loss: 0.0014
Valid set: Average loss: 0.0011, Accuracy: 11608/12000 (96.7333)

Train Epoch: 5  Average loss: 0.0013
Valid set: Average loss: 0.0012, Accuracy: 11589/12000 (96.5750)

Train Epoch: 6  Average loss: 0.0012
Valid set: Average loss: 0.0011, Accuracy: 11618/12000 (96.8167)

Train Epoch: 7  Average loss: 0.0011
Valid set: Average loss: 0.0010, Accuracy: 11614/12000 (96.7833)

Train Epoch: 8  Average loss: 0.0011
Valid set: Average loss: 0.0011, Accuracy: 11639/12000 (96.9917)

Train Epoch: 9  Average loss: 0.0010
Valid set: Average loss: 0.0010, Accuracy: 11645/12000 (97.0417)

Train Epoch: 10  Average loss: 0.0010
Valid set: Average loss: 0.0010, Accuracy: 11660/12000 (97.1667)

Train Epoch: 11  Average loss: 0.0010
Valid set: Average loss: 0.0009, Accuracy: 11692/12000 (97.4333)

Train Epoch: 12  Average loss: 0.0010
Valid set: Average loss: 0.0011, Accuracy: 11647/12000 (97.0583)

Train Epoch: 13  Average loss: 0.0009
Valid set: Average loss: 0.0010, Accuracy: 11670/12000 (97.2500)

Train Epoch: 14  Average loss: 0.0009
Valid set: Average loss: 0.0010, Accuracy: 11678/12000 (97.3167)

Train Epoch: 15  Average loss: 0.0008
Valid set: Average loss: 0.0010, Accuracy: 11693/12000 (97.4417)

Train Epoch: 16  Average loss: 0.0009
Valid set: Average loss: 0.0011, Accuracy: 11642/12000 (97.0167)

Train Epoch: 17  Average loss: 0.0008
Valid set: Average loss: 0.0010, Accuracy: 11672/12000 (97.2667)

Train Epoch: 18  Average loss: 0.0008
Valid set: Average loss: 0.0010, Accuracy: 11686/12000 (97.3833)

Train Epoch: 19  Average loss: 0.0008
Valid set: Average loss: 0.0010, Accuracy: 11696/12000 (97.4667)

Train Epoch: 20  Average loss: 0.0008
Valid set: Average loss: 0.0010, Accuracy: 11680/12000 (97.3333)

============================================

Test set: Average loss: 0.0028, Accuracy: 9758/10000 (97.5800)
(0.002776315849056118, 0.9758)
```

![image](https://user-images.githubusercontent.com/70457814/143507999-54bd4c25-9277-4c22-8b07-3e0e67ce8ad3.png)

**Comment by kaiyun717** [comment]
# 2(c)

```python
model = nn.Sequential(
    # TODO
    nn.Conv2d(1, 32, kernel_size=(5,5), stride=2),   
    nn.ReLU(),
    nn.MaxPool2d(2),
    nn.Dropout(p=0.5),

    nn.Conv2d(32, 64, kernel_size=(2,2)),
    nn.BatchNorm2d(num_features=64),
    nn.ReLU(),
    nn.MaxPool2d(2),
    
    nn.Flatten(),
    nn.Linear(256, 128),

    nn.ReLU(),
    nn.Dropout(p=0.25),

    nn.Linear(128, 10),
    # nn.Softmax()
)
```

![image](https://user-images.githubusercontent.com/70457814/143660612-44f2a433-f0b3-4a4a-9a58-0c3f91313c8e.png)


# FINAL
```python
model = nn.Sequential(
    # TODO
    nn.Conv2d(1, 32, kernel_size=(6,6), stride=2),   
    nn.ReLU(),
    nn.MaxPool2d(2),
    nn.Dropout(p=0.5),

    nn.Conv2d(32, 64, kernel_size=(2,2)),
    nn.BatchNorm2d(num_features=64),
    nn.ReLU(),
    nn.MaxPool2d(2),
    
    nn.Flatten(),
    nn.Linear(256, 128),
    nn.ReLU(),
    nn.Dropout(p=0.25),

    nn.Linear(128, 10)
)
```

![image](https://user-images.githubusercontent.com/70457814/143661198-4dd1d535-f2b7-4c80-bc28-7ec5432b580b.png)

![image](https://user-images.githubusercontent.com/70457814/143661200-7b2413dd-c1cd-4889-9474-37b7fe94bc18.png)

```console
Train Epoch: 1  Average loss: 0.0109
Valid set: Average loss: 0.0030, Accuracy: 11649/12000 (97.0750)

Train Epoch: 2  Average loss: 0.0060
Valid set: Average loss: 0.0029, Accuracy: 11660/12000 (97.1667)

Train Epoch: 3  Average loss: 0.0052
Valid set: Average loss: 0.0026, Accuracy: 11699/12000 (97.4917)

Train Epoch: 4  Average loss: 0.0049
Valid set: Average loss: 0.0022, Accuracy: 11744/12000 (97.8667)

Train Epoch: 5  Average loss: 0.0045
Valid set: Average loss: 0.0020, Accuracy: 11767/12000 (98.0583)

Train Epoch: 6  Average loss: 0.0044
Valid set: Average loss: 0.0020, Accuracy: 11770/12000 (98.0833)

Train Epoch: 7  Average loss: 0.0042
Valid set: Average loss: 0.0019, Accuracy: 11779/12000 (98.1583)

Train Epoch: 8  Average loss: 0.0039
Valid set: Average loss: 0.0019, Accuracy: 11786/12000 (98.2167)

Train Epoch: 9  Average loss: 0.0040
Valid set: Average loss: 0.0017, Accuracy: 11804/12000 (98.3667)

Train Epoch: 10  Average loss: 0.0037
Valid set: Average loss: 0.0017, Accuracy: 11817/12000 (98.4750)

Test set: Average loss: 0.0014, Accuracy: 9859/10000 (98.5900)
(0.0014439356088347267, 0.9859)
```

![image](https://user-images.githubusercontent.com/70457814/143661163-b41266dc-71c3-4d64-b013-0f4fcbcc7e5a.png)

**Comment by kaiyun717** [comment]
# 2(a)
- Image Shape: (#channels, height, width) = (1, 28, 28)
- Number of Training Images: 48,000
- Number of Testing Images: 10,000

# 2(b)

## Final Model
```python
model = nn.Sequential(
    nn.Flatten(),
    nn.Linear(28*28, 128),
    nn.ReLU(),
    nn.Dropout(p=0.30),
    nn.Linear(128, 64),
    nn.ReLU(),
    nn.Linear(64, 10)
)

num_hidden_units = 128 + 64
batch_size = 100
n_epochs = 20
lr = 5e-3
```

## Results
Test Accuracy: 97.58%
![image](https://user-images.githubusercontent.com/70457814/143791816-02ee15ae-4595-4070-8b7e-57f94642c79a.png)

## Trials and Errors
- At first, I tried using Softmax layer after the second hidden layer. However, when I only used ReLU layers for activation, the results were much better. For example, for the exact same models, changing Softmax layer to a ReLU layer after the second hidden layer improved test accuracy from 96.44% to 97.42%.
- I tried a learning rate of 1e-2, but that was too big for stable convergence. However, 1e-3 was too small. Thus, I chose 5e-3.
- For batch_size, I tried 32, 64, 128, but settled down with 100 since 128 seemed to be too big.
- Increasing Dropout probability also helped. I first started with 0.1 after both hidden layers. However, having one Dropout layer of probability 0.3 turned out to be the best.
- I tried my best to achieve a test accuracy of over 97.5% with 10 epochs, but eventually used 20 epochs. With the exact same model, I was able to achieve 97.49% test accuracy after 10 epochs. Thus, I added 10 additional epochs in the next (and final) training incident.

---

# 2(c)

## Final Model
```python
model = nn.Sequential(
    nn.Conv2d(1, 32, kernel_size=(6,6), stride=2),   
    nn.ReLU(),
    nn.MaxPool2d(2),
    nn.Dropout(p=0.5),

    nn.Conv2d(32, 64, kernel_size=(2,2)),
    nn.BatchNorm2d(num_features=64),
    nn.ReLU(),
    nn.MaxPool2d(2),
    
    nn.Flatten(),
    nn.Linear(256, 128),
    nn.ReLU(),
    nn.Dropout(p=0.25),

    nn.Linear(128, 10)
)

num_hidden_units = 43_754
batch_size = 32
n_epochs = 10
lr = 5e-3
```

## Results
Test Accuracy: 98.54%
![image](https://user-images.githubusercontent.com/70457814/143792295-fb9b340f-cd64-468f-8ca7-8779073a88fc.png)

## Trials and Errors
- I first tried without using `BatchNorm2d` and the test accuracy was 95.74%.
- Having a bigger kernel size for the first convolutional layer results in better test accuracy. In such a case, it's good to have a stride of 2 to reduce the output dimension.
- Again, using a Softmax layer actually hurts the performance for some reason. I think it's because Softmax limits the output value to be between 0 and 1. Thus, limiting the size of gradient values in back-propagation. In many cases, this may result in the "vanishing gradients" issue.
- Having one Dropout layer after the first convolutional layer, and another after the hidden layer definitely helps the model's performance. 

## Regularization Method
- Definitely having at least two Dropout layers with reasonably high dropout probabilities help.
- As the *Hint* suggests, Batch Norm is most effective after both convolutional layers, but before activation or MaxPool2d. 

---

# 3(a)

## Final Model
```python
class SingleHeadAttention(nn.Module):
    def __init__(self, input_dim, inner_dim, dropout = 0.):
        """
        inner_dim: d_k, the query dimension.
        """
        super().__init__()
        self.input_dim = input_dim
        self.inner_dim = inner_dim
        self.dropout_rate = dropout
        self.scale = inner_dim ** 0.5
        
        self.key_layer = nn.Linear(input_dim, inner_dim)
        self.query_layer = nn.Linear(input_dim, inner_dim)
        self.value_layer = nn.Linear(input_dim, inner_dim)

        self.softmax = nn.Softmax(dim=-1)
        self.dropout = nn.Dropout(p=dropout)

    def forward(self, x):
        key = self.dropout(self.key_layer(x))
        query = self.dropout(self.query_layer(x))
        value = self.dropout(self.value_layer(x))

        q_k = torch.matmul(query, torch.transpose(key, 1, 2))
        softmax = self.softmax(q_k / self.scale)

        out = torch.matmul(softmax, value)
        return out
```

# 3(b)
```python
class ViTLayer(nn.Module):
    def __init__(self, dim, heads, dim_head, mlp_dim, dropout = 0.):
        super().__init__()
        self.attn = nn.Sequential(
            nn.LayerNorm(dim),
            Attention(dim=dim, heads=heads, dim_head=dim_head, dropout=dropout),
            nn.Dropout(p=dropout),

            nn.Linear(heads*dim_head, dim)
        )
        self.feedforward = nn.Sequential(
            nn.LayerNorm(dim),
            nn.Linear(dim, mlp_dim),
            nn.ReLU(),
            nn.Dropout(p=dropout),
            
            nn.Linear(mlp_dim, dim),
            nn.ReLU(),
            nn.Dropout(p=dropout)
        )

    def forward(self, x):
        attention = self.attn(x)
        residual = torch.add(attention, x)
        return torch.add(residual, self.feedforward(residual))
```

# 3(c)
Test Accuracy: 97.89%
![image](https://user-images.githubusercontent.com/70457814/143793059-9d4a456b-1e0c-48f5-85c7-7198c5211006.png)
